### PR TITLE
Fix static build on Windows

### PR DESCRIPTION
**Pull request by isaevil**
Signed-off-by: Isaev, Ilya <ilya.isaev@intel.com>

### CODE REVIEW

File: src/tbb/governor.cpp [hunk 1]
```diff
@@ -167,8 +167,20 @@ static std::uintptr_t get_stack_base(std::size_t stack_size) {
 #endif /* USE_PTHREAD */
 }
 
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+struct thread_destructor {
+   ~thread_destructor() {
+       governor::terminate_external_thread();
+   }
+}
```
Comment (pavelkumbrasev) [inline src/tbb/governor.cpp:175]:
```suggestion
}
```

Comment (alexey-katranov) [reply src/tbb/governor.cpp:175]:
```suggestion
};
```

File: src/tbb/governor.cpp [hunk 2]
```diff
@@ -167,8 +167,20 @@ static std::uintptr_t get_stack_base(std::size_t stack_size) {
...
+   }
+} 
+static thread_local thr_destructor;
+#endif // (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+
 void governor::init_external_thread() {
     one_time_init();
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+    (void)thr_destructor;
+#endif // (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
```
Comment (alexey-katranov) [inline src/tbb/governor.cpp:183]: Move it after ` set_thread_data(td);` on line 206. Consider introducing some function like `register_external_thread_destructor` and moving the whole implementation into this function.

File: test/tbb/test_global_control.cpp [hunk 3]
```diff
@@ -222,6 +222,19 @@ void TestTerminationAndAutoinit(bool autoinit) {
...
 }
 
+TEST_CASE("test decrease reference") {
+    auto thread_func = [] {
+        tbb::parallel_for(0, 1, [](int) {});
+    };
+
+    tbb::task_scheduler_handle handle = tbb::task_scheduler_handle::get();
+
+    std::thread thr(thread_func);
```
Comment (alexey-katranov) [inline test/tbb/test_global_control.cpp:232]:
```suggestion
    tbb::task_scheduler_handle handle = tbb::task_scheduler_handle::get();

    std::thread thr([] { tbb::parallel_for(0, 1, [](int) {}); });
```

File: test/tbb/test_global_control.cpp [hunk 4]
```diff
@@ -222,6 +222,19 @@ void TestTerminationAndAutoinit(bool autoinit) {
     REQUIRE(res2);
 }
 
+TEST_CASE("test decrease reference") {
```
Comment (alexey-katranov) [inline test/tbb/test_global_control.cpp:225]:
```suggestion
//! Check no reference leak for an external thread
//! \brief \ref regression \ref error_guessing
TEST_CASE("test decrease reference") {
```

File: src/tbb/governor.h [hunk 5]
```diff
@@ -87,6 +87,10 @@ class governor {
         The auto_init argument specifies if the call is due to automatic initialization. **/
     static void init_external_thread();
 
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+    static void register_external_thread_destructor();
```
Comment (pavelkumbrasev) [inline src/tbb/governor.h:91]: Maybe just add it as a standalone method in governor.cpp?

File: src/tbb/governor.cpp [hunk 6]
```diff
@@ -192,6 +203,9 @@ void governor::init_external_thread() {
     td.my_arena_slot->occupy();
     a.my_market->add_external_thread(td);
     set_thread_data(td);
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+    register_external_thread_destructor();
+#endif // (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
```
Comment (pavelkumbrasev) [inline src/tbb/governor.cpp:208]: It's just one line)
```suggestion
#endif
```

File: src/tbb/governor.cpp [hunk 7]
```diff
@@ -167,6 +167,17 @@ static std::uintptr_t get_stack_base(std::size_t stack_size) {
 #endif /* USE_PTHREAD */
 }
 
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+void governor::register_external_thread_destructor() {
+    struct thread_destructor {
```
Comment (pavelkumbrasev) [inline src/tbb/governor.cpp:172]: `thread_data_destructor` ?

File: src/tbb/governor.cpp [hunk 8]
```diff
@@ -168,13 +168,13 @@ static std::uintptr_t get_stack_base(std::size_t stack_size) {
...
-    struct thread_destructor {
-        ~thread_destructor() {
+static void register_external_thread_destructor() {
+    struct thread_data_destructor {
+        ~thread_data_destructor() {
             governor::terminate_external_thread();
         }
     };
-    static thread_local thread_destructor thr_destructor;
+    static thread_local thread_data_destructor thr_destructor;
```
Comment (alexey-katranov) [inline src/tbb/governor.cpp:177]:
```suggestion
    // ~thread_data_destructor() will be call during the calling thread termination
    static thread_local thread_data_destructor thr_destructor;
```

File: src/tbb/governor.cpp [hunk 9]
```diff
@@ -192,6 +204,9 @@ void governor::init_external_thread() {
     td.my_arena_slot->occupy();
     a.my_market->add_external_thread(td);
     set_thread_data(td);
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+    register_external_thread_destructor();
```
Comment (alexey-katranov) [inline src/tbb/governor.cpp:208]:
```suggestion
    // The external thread destructor is called from dllMain but it is not available with a static build.
    // Therefore, we need to register the current thread to call the destructor during thread termination.
    register_external_thread_destructor();
```

File: src/tbb/governor.cpp [hunk 10]
```diff
@@ -167,6 +167,18 @@ static std::uintptr_t get_stack_base(std::size_t stack_size) {
...
 
+#if (_WIN32||_WIN64) && !__TBB_DYNAMIC_LOAD_ENABLED
+static void register_external_thread_destructor() {
+    struct thread_data_destructor {
+        ~thread_data_destructor() {
+            governor::terminate_external_thread();
+        }
+    };
+    // ~thread_data_destructor() will be call during the calling thread termination
+    static thread_local thread_data_destructor thr_destructor;
```
Comment (alexey-katranov) [inline src/tbb/governor.cpp:178]:
```suggestion
    struct thread_destructor {
        ~thread_destructor() {
            governor::terminate_external_thread();
        }
    };
    // ~thread_destructor() will be call during the calling thread termination
    static thread_local thread_destructor thr_destructor;
```

**Review by alexey-katranov — APPROVED** [review]

**Pull request merged**